# DSO 105 Intermediate Stats L3 - Python

In [1]:
## When you have 1 or more categorical varialbe you can either do: proportion testing or Chi-Square

In [2]:
#### Page 3

By the end of the lesson, you should be able to conduct the following in both Python and R:

* One proportion tests

* Two proportion tests

* Independent Chi-Squares

* Goodness of fit Chi-Squares

* McNemar Chi-Squares

In addition, you should also be able to:

* Differentiate between the different types of categorical tests

* Test assumptions and perform post hoc analysis for Chi-Squares in R

In [1]:
## import packages

import pandas as pd
import scipy, scipy.stats
import numpy as np
from statsmodels.stats.proportion import proportions_ztest
import statsmodels as sm
from statsmodels.stats.contingency_tables import mcnemar

## One Proportion Testing in Python

Now that you know how to do one proportion testing in R, you will try the exact same problem in Python, 

examining the proportion of jelly beans to chocolate eggs in your Easter basket when you have 15 jelly beans out of a total 43 pieces of candy.

You'll start by importing a few packages. You'll need numpy as well as some specifics from statsmodels to do your one proportion z test.

Then you can define your count as the number of jelly beans, 15, and your nobs, standing for the number of observations, as the total number of pieces of candy, which is 43. 

You will also define the proportion value to which you want to compare. 

Since you are trying to test whether the number of jelly beans is equal to the number of chocolate eggs, you will use a proportion of .5, which would be half and each would be equal. 

Lastly you will define the output of your function proportions_ztest() as stat and pval, and will feed into the proportions_ztest() function the arguments you just defined above: count, nobs, and value.

In [4]:
count = 15
nobs = 43
value = .5

In [5]:
stat, pval = proportions_ztest(count, nobs, value)

In [6]:
print(stat, pval)

-2.079806538622099 0.03754328113448803


Finally, go ahead and print your results out so you can revel in their glory! What a surprise - you have come up with the exact same answer in both R and Python!

In [7]:
## but it is NOT the same, this p-value is a little higher

In [8]:
## the 2nd question in the quiz is also contradictory to what the curriculum says. what is nobs?

In [9]:
#### Page 5

## Two Proportion Testing

You will use a two proportion z test when you want to compare the proportions of two different categories to the whole.

## Two Proportion Testing in Python

In order to complete two proportion testing in Python, you will use the same function, but will need to handle it a bit differently.

For the Easter basket scenario from the last page, in which there are 15 jelly beans and 28 chocolate eggs, with 7 pink jelly beans and 12 pink chocolate eggs, here's what the Python code would look like:

In [10]:
stat, pval = proportions_ztest([7, 12], [15, 28])
print(stat,pval)

0.23974366706563624 0.810528980523634


Just like with R, you have found that this is not a significant difference in the proportions of pink in either type of candy.

In [11]:
## ??

In [12]:
#### Page 8

## Goodness of Fit Chi-Squares in Python

Running a goodness of fit Chi-Square in Python is very similar to running a goodness of fit Chi-Square in R.

## Load in Data

You'll work on the same Star Wars survey as you did in R.

In [2]:
SW = pd.read_csv('../../datasets/SW_survey_renamed.csv')

In [3]:
SW.head()

RespondentID SeenYN FanYN                                   SeenIYN  \
0    3292879998    Yes   Yes  Star Wars: Episode I  The Phantom Menace   
1    3292879538     No   NaN                                       NaN   
2    3292765271    Yes    No  Star Wars: Episode I  The Phantom Menace   
3    3292763116    Yes   Yes  Star Wars: Episode I  The Phantom Menace   
4    3292731220    Yes   Yes  Star Wars: Episode I  The Phantom Menace   

                                      SeenIIYN  \
0  Star Wars: Episode II  Attack of the Clones   
1                                          NaN   
2  Star Wars: Episode II  Attack of the Clones   
3  Star Wars: Episode II  Attack of the Clones   
4  Star Wars: Episode II  Attack of the Clones   

                                     SeenIIIYN  \
0  Star Wars: Episode III  Revenge of the Sith   
1                                          NaN   
2  Star Wars: Episode III  Revenge of the Sith   
3  Star Wars: Episode III  Revenge of the Sith   
4  Star Wars: Episode III  Revenge of the Sith   

                            SeenIVYN  \
0  Star Wars: Episode IV  A New Hope   
1                                NaN   
2                                NaN   
3  Star Wars: Episode IV  A New Hope   
4  Star Wars: Episode IV  A New Hope   

                                        SeenVYN  \
0  Star Wars: Episode V The Empire Strikes Back   
1                                           NaN   
2                                           NaN   
3  Star Wars: Episode V The Empire Strikes Back   
4  Star Wars: Episode V The Empire Strikes Back   

                                   SeenVIYN  RankI  ...      Favorable_Yoda  \
0  Star Wars: Episode VI Return of the Jedi    3.0  ...      Very favorably   
1                                       NaN    NaN  ...                 NaN   
2                                       NaN    1.0  ...    Unfamiliar (N/A)   
3  Star Wars: Episode VI Return of the Jedi    5.0  ...      Very favorably   
4  Star Wars: Episode VI Return of the Jedi    5.0  ...  Somewhat favorably   

                          ShotFirst  ExpandedUniverseYN  \
0  I don't understand this question                 Yes   
1                               NaN                 NaN   
2  I don't understand this question                  No   
3  I don't understand this question                  No   
4                            Greedo                 Yes   

   FanExpandedUniverseYN  StarTrekFanYN Gender    Age     Household Income  \
0                     No             No   Male  18-29                  NaN   
1                    NaN            Yes   Male  18-29         $0 - $24,999   
2                    NaN             No   Male  18-29         $0 - $24,999   
3                    NaN            Yes   Male  18-29  $100,000 - $149,999   
4                     No             No   Male  18-29  $100,000 - $149,999   

                          Education            Location  
0                High school degree      South Atlantic  
1                   Bachelor degree  West South Central  
2                High school degree  West North Central  
3  Some college or Associate degree  West North Central  
4  Some college or Associate degree  West North Central  

[5 rows x 38 columns]

## Question Set Up

You will be testing the same premise as when you did a goodness of fit Chi-Square in R: You found something online that mentioned that 90% of people are Star Wars fans, and you want to see if that holds true in your own survey. 
    
In this way, you are comparing your sample (the survey) to the population at large (what you read online).

## Data Wrangling

You will need to get the number of people who were and were not fans of Star Wars. Luckily, this is relatively easy to do with the pandas function value_counts():

In [15]:
SW.FanYN.value_counts()

Yes    552
No     284
Name: FanYN, dtype: int64

## Run the Analysis

Now you are ready to run your analysis! You will first create a variable that houses the observed values:

In [16]:
observed_values = np.array([552, 284])

Then you will create a variable that houses the expected values. Unlike R, Python requires you to have raw numbers, not percentages here, so you will ned to calculate the values yourself. First, add up your expected values to get the total: 552 + 284 = 836. Then, multiply that number by .9 to get what percentage would be 90%. The number is 752 - this becomes your first expected value. Then subtract hat number, 752, from the total, and you will get your other value: 836-752 = 84.

In [17]:
expected_values = np.array([752, 84])

Once you have those two variables, it is simply a matter of plugging them into your chisquare() function that comes in scipy.stats:

In [18]:
scipy.stats.chisquare(observed_values, f_exp=expected_values)

Power_divergenceResult(statistic=529.3819655521784, pvalue=3.849512370977756e-117)

And the output is provided below. The one labeled statistic is your Chi-Square value, and the one labeled pvalue is - suprise, surprise~ - your p value!

It will most likely be written in scientific notation. It is in this case, and so this means that this value is extremely significant - you are moving the decimal over to the left 117 times, which means that there are 116 zeros in front of that 3! And remember that a significant goodness of fit Chi-Square means that your sample significantly differs from the population in some way.

In [19]:
## the population was the internet survery of 90/100 percent Star Wars fans

In [20]:
#### Page 10

## McNemar Chi-Square Python

You will now learn how to conduct McNemar Chi-Squares in Python. The process is much the same, but the output leaves something to be desired.

## Load in Data

You will be using the same bakery data as you did with R. It has the date and time of each bakery item sold.

In [4]:
bakerySales = pd.read_csv('../../datasets/bakery_sales.csv')

In [5]:
bakerySales.head()

Date         Time  Transaction           Item
0  10/30/2016   9:58:11 AM            1          Bread
1  10/30/2016  10:05:34 AM            2   Scandinavian
2  10/30/2016  10:05:34 AM            2   Scandinavian
3  10/30/2016  10:07:57 AM            3  Hot chocolate
4  10/30/2016  10:07:57 AM            3            Jam

## Question Set Up

You will be answering the following question:

Do the sales of coffee change from the beginning of the month to the end of the month?

## Data Wrangling

Just like with R, you will need to do some data wrangling.

## Separating the Pieces of the Date Variable

The first order of business is to separate out your Date column. You can do this with the function str.split():

In [23]:
bakerySales1 = bakerySales['Date'].str.split('/', expand=True).rename(columns = lambda x: "Date" + str(x +1))

And then of course you'll need to put your data back together again:

In [24]:
bakerySales2 = pd.concat([bakerySales, bakerySales1], axis=1)

In [25]:
bakerySales2.head()

Date         Time  Transaction           Item Date1 Date2 Date3
0  10/30/2016   9:58:11 AM            1          Bread    10    30  2016
1  10/30/2016  10:05:34 AM            2   Scandinavian    10    30  2016
2  10/30/2016  10:05:34 AM            2   Scandinavian    10    30  2016
3  10/30/2016  10:07:57 AM            3  Hot chocolate    10    30  2016
4  10/30/2016  10:07:57 AM            3            Jam    10    30  2016

## Changing Day to an Integer

Next you'll need to recode the Date2 variable so that it provides information about beginning or ending of the month. To do this, your Date2 variable will need to be an integer. 

You can double check that it is with the function info():

In [26]:
bakerySales2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21293 entries, 0 to 21292
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Date         21293 non-null  object
 1   Time         21293 non-null  object
 2   Transaction  21293 non-null  int64 
 3   Item         21293 non-null  object
 4   Date1        21293 non-null  object
 5   Date2        21293 non-null  object
 6   Date3        21293 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.1+ MB


So it looks like Date2 is currently string data, which is common after doing the str.split() function - after all, it literally translates into "string split!" However, this is an easy fix - you can use the astype(int) function:

In [27]:
bakerySales2.Date2 = bakerySales2.Date2.astype(int)

And now if you run info() again, you will find that Date2 is now an integer!

In [28]:
bakerySales2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21293 entries, 0 to 21292
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Date         21293 non-null  object
 1   Time         21293 non-null  object
 2   Transaction  21293 non-null  int64 
 3   Item         21293 non-null  object
 4   Date1        21293 non-null  object
 5   Date2        21293 non-null  int32 
 6   Date3        21293 non-null  object
dtypes: int32(1), int64(1), object(5)
memory usage: 1.1+ MB


## Recoding to Beginning or End of Month

Now that your variable is numeric, you are good to do a recode using the greater than and less than operands. 

You can recode using a function with some if statements, and then apply that function to your data:

In [29]:
def month (series): 
    if series <= 15: 
        return 0
    if series > 16: 
        return 1
bakerySales2['DayR'] = bakerySales2["Date2"].apply(month)

## Recoding to Coffee or Other

Next, you will recode the Item variable into something that is Coffee or Not Coffee. You will use the same format as the recode above:

In [30]:
def coff (series):
    if series == "Coffee":
        return 1
    if series != "Coffee":
        return 2
bakerySales2['CoffeeYN'] = bakerySales2['Item'].apply(coff)

## Make a Contingency Table


Next, you will need to make a contingency table, since the function for McNemar Chi-Squares in Python will not accept raw data. 

Happily, the pd.crosstab() function you learned earlier will do this job easily for you:

In [31]:
bakery_crosstab = pd.crosstab(bakerySales2['DayR'], bakerySales2['CoffeeYN'])
bakery_crosstab

CoffeeYN     1     2
DayR                
0.0       2841  8238
1.0       2491  7126

## Test Assumptions and Run Analyses

In Python, there is no way to test the assumption of at least five expected per cell, which means that if you are analyzing high stakes data, 

where accuracy really matters, Python is NOT the program for you to run a McNemar Chi-Square in.

You will use the function sm.stats.contingency_tables.mcnemar() to run your McNemar Chi-Square. 

It takes the arguments of the crosstab you just created, exact=, which you can set to False, and correction=, which will be set to True.

In [32]:
result = sm.stats.contingency_tables.mcnemar(bakery_crosstab, exact=False, correction=True)

If you just run the code above, you may end up confused - nothing happened! That's because this particular function requires you to actually print your results out yourself:

In [33]:
print(result)

pvalue      0.0
statistic   2839.0003519887364


## Interpret Results

Alright! You now have results, and they are significant - the p value is less than .05, so it looks like different amounts of coffee is sold in the morning and afternoon! 

How does it differ? With Python, you'll NEVER KNOW! It does not provide the ability to look at standardized residuals, so you can't look at post hocs.

## Summary

Although continuous data is usually easier to work with and you can extract more data from it, 

there will be times when you come up against a large pile of categorical data (especially if your company collected it themselves!) 

and you will need some advanced categorical analysis tools in your arsenal! 

You will use one proportion testing when you are comparing the rate of one item to a gold standard rate. 

Two proportion testing will be used to compare rates between items to a gold standard rate. 

Goodness of fit Chi-Squares are used to test whether your sample data could feasibly come from the population as a whole, 

and you can use a McNemar Chi-Square to look at anything that is repeatedly measured that has two categorical variables.

In [ ]:
#### Page 11 (Key Terms)